components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java

```diff
@@ -2147,29 +2236,89 @@ public void updateApplication(Application application) throws APIManagementExcep
+
+            workflowExtRef = apiMgtDAO.getExternalWorkflowReferenceByApplicationID(application.getId());
+
+            ApplicationWorkflowDTO workflowDTO = new ApplicationWorkflowDTO();
+            workflowDTO.setApplication(application);
+            workflowDTO.setWorkflowType(WorkflowConstants.WF_TYPE_AM_APPLICATION_DELETION);
+            workflowDTO.setTenantDomain(tenantDomain);
+            workflowDTO.setTenantId(tenantId);
+            workflowDTO.setExternalWorkflowReference(workflowExtRef);
+
+            // Remove from cache first since we won't be able to find active access tokens
+            // once the application is removed.
             invalidateCachedKeys(application.getId());
-        }catch(APIManagementException ignore){
-            //Log and proceed since we do not want to halt the application delete due to cache invalidation failures.
-            log.warn("Failed to invalidate Gateway Cache " + ignore.getMessage());
+
+            // clean up pending subscription tasks
+            Set<Integer> pendingSubscriptions = apiMgtDAO.getPendingSubscriptionsByApplicationId(applicationId);
+            for (int subscription : pendingSubscriptions) {
+                workflowExtRef = apiMgtDAO.getExternalWorkflowReferenceForSubscription(subscription);
+                createSubscriptionWFExecutor.cleanUpPendingTask(workflowExtRef);
+            }
+
+            // cleanup pending application registration tasks
+            String productionKeyStatus = apiMgtDAO.getRegistrationApprovalState(applicationId, APIConstants
+                    .API_KEY_TYPE_PRODUCTION);
+            String sandboxKeyStatus = apiMgtDAO.getRegistrationApprovalState(applicationId, APIConstants
+                    .API_KEY_TYPE_SANDBOX);
+            if (WorkflowStatus.CREATED.toString().equals(productionKeyStatus)) {
+                workflowExtRef = apiMgtDAO.getRegistrationWFReference(applicationId, APIConstants
+                        .API_KEY_TYPE_PRODUCTION);
+                createProductionRegistrationWFExecutor.cleanUpPendingTask(workflowExtRef);
+            }
+            if (WorkflowStatus.CREATED.toString().equals(sandboxKeyStatus)) {
+                workflowExtRef = apiMgtDAO.getRegistrationWFReference(applicationId, APIConstants.API_KEY_TYPE_SANDBOX);
+                createSandboxRegistrationWFExecutor.cleanUpPendingTask(workflowExtRef);
+            }
+            createApplicationWFExecutor.cleanUpPendingTask(workflowExtRef);
+            removeApplicationWFExecutor.execute(workflowDTO);
+        } catch (WorkflowException e) {
+            String errorMsg = "Could not execute Workflow, " + WorkflowConstants.WF_TYPE_AM_APPLICATION_DELETION + " " +
+                    "for applicationID " + application.getId();
+            handleException(errorMsg, e);
+        } finally {
+            if (isTenantFlowStarted) {
+                PrivilegedCarbonContext.endTenantFlow();
```

Comment:
I agreed with what @prabathabey and @imesh mentioned in their above comments. I don't see any particular need to check whether a tenantFlow is started or not. Could you explain the reason for this check?